storage/grpc_reader_multi_range.go

```diff
@@ -38,3 +51,834 @@ func (g *readIDGenerator) Next() int64 {
+	}
+
+	b := bucketResourceName(globalProjectAlias, params.bucket)
+	readSpec := &storagepb.BidiReadObjectSpec{
+		Bucket:                    b,
+		Object:                    params.object,
+		CommonObjectRequestParams: toProtoCommonObjectRequestParams(params.encryptionKey),
+	}
+	if params.gen >= 0 {
+		readSpec.Generation = params.gen
+	}
+	if params.handle != nil && len(*params.handle) > 0 {
+		readSpec.ReadHandle = &storagepb.BidiReadHandle{
+			Handle: *params.handle,
+		}
+	}
+
+	mCtx, cancel := context.WithCancel(ctx)
+
+	// Create the manager
+	manager := newMultiRangeDownloaderManager(mCtx, c, s, params, readSpec, cancel, ctx)
+
+	mrd := &MultiRangeDownloader{
+		impl:    manager,
+		spanCtx: ctx,
+	}
+	manager.publicMRD = mrd
+
+	manager.wg.Add(1)
+	go func() {
+		defer manager.wg.Done()
+		manager.eventLoop()
+	}()
+
+	// Wait for attributes to be ready
+	select {
+	case <-manager.attrsReady:
+		if manager.permanentErr != nil {
+			cancel()
+			manager.wg.Wait()
+			return nil, manager.permanentErr
+		}
+		mrd.Attrs = manager.attrs
+		return mrd, nil
+	case <-ctx.Done():
+		cancel()
+		manager.wg.Wait()
+		return nil, ctx.Err()
+	}
+}
+
+// --- mrdCommand Interface and Implementations ---
+type mrdCommand interface {
+	apply(ctx context.Context, m *multiRangeDownloaderManager)
+}
+type mrdAddCmd struct {
+	output   io.Writer
+	offset   int64
+	length   int64
+	callback func(int64, int64, error)
+}
+
+func (c *mrdAddCmd) apply(ctx context.Context, m *multiRangeDownloaderManager) {
+	m.handleAddCmd(ctx, c)
+}
+
+type mrdCloseCmd struct {
+	err error
+}
+
+func (c *mrdCloseCmd) apply(ctx context.Context, m *multiRangeDownloaderManager) {
+	m.handleCloseCmd(ctx, c)
+}
+
+type mrdWaitCmd struct {
+	doneC chan struct{}
+}
+
+func (c *mrdWaitCmd) apply(ctx context.Context, m *multiRangeDownloaderManager) {
+	m.handleWaitCmd(ctx, c)
+}
+
+type mrdGetHandleCmd struct {
+	respC chan []byte
+}
+
+func (c *mrdGetHandleCmd) apply(ctx context.Context, m *multiRangeDownloaderManager) {
+	select {
+	case <-m.attrsReady:
+		select {
+		case c.respC <- m.lastReadHandle:
+		case <-m.ctx.Done():
+			close(c.respC)
+		}
+	case <-m.ctx.Done():
+		close(c.respC)
+	}
+}
+
+type mrdErrorCmd struct {
+	respC chan error
+}
+
+func (c *mrdErrorCmd) apply(ctx context.Context, m *multiRangeDownloaderManager) {
+	select {
+	case c.respC <- m.permanentErr:
+	case <-ctx.Done():
+		close(c.respC)
+	}
+}
+
+// --- mrdSessionResult ---
+type mrdSessionResult struct {
+	resp     *storagepb.BidiReadObjectResponse
+	err      error
+	redirect *storagepb.BidiReadObjectRedirectedError
+}
+
+var errClosed = errors.New("downloader closed")
+
+// --- multiRangeDownloaderManager ---
+
+type multiRangeDownloaderManager struct {
+	ctx          context.Context
+	cancel       context.CancelFunc
+	client       *grpcStorageClient
+	settings     *settings
+	params       *newMultiRangeDownloaderParams
+	wg           sync.WaitGroup
+	cmdC         chan mrdCommand
```

Comment:
I generally prefer for channels to use the plural of the contents, unless already taken by a different variable:
```suggestion
	cmds         chan mrdCommand
```